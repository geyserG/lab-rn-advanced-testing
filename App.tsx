import React, {useState} from 'react';
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  SafeAreaView,
} from 'react-native';

function calculateWinner(squares: string[]) {
  const lines = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8],
    [0, 3, 6],
    [1, 4, 7],
    [2, 5, 8],
    [0, 4, 8],
    [2, 4, 6],
  ];
  for (let i = 0; i < lines.length; i++) {
    const [a, b, c] = lines[i];
    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
      return squares[a];
    }
  }
  return null;
}

const App = () => {
  const [history, setHistory] = useState([Array(9).fill(null)]);
  const [currentMove, setCurrentMove] = useState(0);
  const xIsNext = currentMove % 2 === 0;
  const currentSquares = history[currentMove];

  function handlePlay(nextSquares: string[]) {
    const nextHistory = [...history.slice(0, currentMove + 1), nextSquares];
    setHistory(nextHistory);
    setCurrentMove(nextHistory.length - 1);
  }

  return (
    <SafeAreaView>
      <View style={styles.container}>
        <View style={styles.gameBoard}>
          <Board
            xIsNext={xIsNext}
            squares={currentSquares}
            onPlay={handlePlay}
          />
        </View>
      </View>
    </SafeAreaView>
  );
};

type SquareProps = {
  value: string;
  onPress: () => void;
};

function Square({value, onPress}: SquareProps) {
  return (
    <TouchableOpacity testID="square" style={styles.square} onPress={onPress}>
      <Text>{value}</Text>
    </TouchableOpacity>
  );
}

type BoardProps = {
  xIsNext: boolean;
  squares: string[];
  onPlay: (nextSquares: string[]) => void;
};

const Board = ({xIsNext, squares, onPlay}: BoardProps) => {
  function handleSquarePress(i: number) {
    if (calculateWinner(squares) || squares[i]) {
      return;
    }
    const nextSquares = squares.slice();
    if (xIsNext) {
      nextSquares[i] = 'X';
    } else {
      nextSquares[i] = 'O';
    }
    onPlay(nextSquares);
  }

  const winner = calculateWinner(squares);
  let status;
  if (winner) {
    status = 'Winner: ' + winner;
  } else {
    status = 'Next player: ' + (xIsNext ? 'X' : 'O');
  }

  return (
    <View testID="board" style={styles.board}>
      <Text testID="status" style={styles.status}>
        {status}
      </Text>
      <View style={styles.row}>
        <Square value={squares[0]} onPress={() => handleSquarePress(0)} />
        <Square value={squares[1]} onPress={() => handleSquarePress(1)} />
        <Square value={squares[2]} onPress={() => handleSquarePress(2)} />
      </View>
      <View style={styles.row}>
        <Square value={squares[3]} onPress={() => handleSquarePress(3)} />
        <Square value={squares[4]} onPress={() => handleSquarePress(4)} />
        <Square value={squares[5]} onPress={() => handleSquarePress(5)} />
      </View>
      <View style={styles.row}>
        <Square value={squares[6]} onPress={() => handleSquarePress(6)} />
        <Square value={squares[7]} onPress={() => handleSquarePress(7)} />
        <Square value={squares[8]} onPress={() => handleSquarePress(8)} />
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {},
  gameBoard: {},
  board: {},
  status: {
    fontSize: 16,
    color: '#333',
    paddingVertical: 10,
  },
  row: {
    flexDirection: 'row',
  },
  square: {
    width: 50,
    height: 50,
    borderWidth: 1,
    margin: 2,
    borderColor: '#666',
    justifyContent: 'center',
    alignItems: 'center',
  },
});

export default App;
