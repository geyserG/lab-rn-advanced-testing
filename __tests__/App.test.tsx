import React from 'react';
import {render, fireEvent, waitFor} from '@testing-library/react-native';
import App from '../App';

describe('App', () => {
  it('should render the game board', () => {
    const {getByTestId} = render(<App />);
    const board = getByTestId('board');
    expect(board).toBeTruthy();
    expect(board.children.length).toBe(4);
  });

  it('should render the status text', () => {
    const {getByTestId} = render(<App />);
    const status = getByTestId('status');
    expect(status).toBeTruthy();

    expect(status.props.children).toBe('Next player: X');
  });

  it('should update the game state when a cell is tapped', () => {
    const {getAllByTestId, getByTestId} = render(<App />);
    const squares = getAllByTestId('square');
    fireEvent.press(squares[0]);
    waitFor(() => {
      expect(getByTestId('status').props.children).toBe('Next player: O');
    });
  });
});