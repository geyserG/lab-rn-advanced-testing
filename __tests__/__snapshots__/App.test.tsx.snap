// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<RCTSafeAreaView>
  <View
    style={{}}
  >
    <View
      style={{}}
    >
      <View
        style={{}}
        testID="board"
      >
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 16,
              "paddingVertical": 10,
            }
          }
          testID="status"
        >
          Next player: X
        </Text>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderColor": "#666",
                "borderWidth": 1,
                "height": 50,
                "justifyContent": "center",
                "margin": 2,
                "opacity": 1,
                "width": 50,
              }
            }
            testID="square"
          >
            <Text />
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderColor": "#666",
                "borderWidth": 1,
                "height": 50,
                "justifyContent": "center",
                "margin": 2,
                "opacity": 1,
                "width": 50,
              }
            }
            testID="square"
          >
            <Text />
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderColor": "#666",
                "borderWidth": 1,
                "height": 50,
                "justifyContent": "center",
                "margin": 2,
                "opacity": 1,
                "width": 50,
              }
            }
            testID="square"
          >
            <Text />
          </View>
        </View>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderColor": "#666",
                "borderWidth": 1,
                "height": 50,
                "justifyContent": "center",
                "margin": 2,
                "opacity": 1,
                "width": 50,
              }
            }
            testID="square"
          >
            <Text />
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderColor": "#666",
                "borderWidth": 1,
                "height": 50,
                "justifyContent": "center",
                "margin": 2,
                "opacity": 1,
                "width": 50,
              }
            }
            testID="square"
          >
            <Text />
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderColor": "#666",
                "borderWidth": 1,
                "height": 50,
                "justifyContent": "center",
                "margin": 2,
                "opacity": 1,
                "width": 50,
              }
            }
            testID="square"
          >
            <Text />
          </View>
        </View>
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderColor": "#666",
                "borderWidth": 1,
                "height": 50,
                "justifyContent": "center",
                "margin": 2,
                "opacity": 1,
                "width": 50,
              }
            }
            testID="square"
          >
            <Text />
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderColor": "#666",
                "borderWidth": 1,
                "height": 50,
                "justifyContent": "center",
                "margin": 2,
                "opacity": 1,
                "width": 50,
              }
            }
            testID="square"
          >
            <Text />
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "borderColor": "#666",
                "borderWidth": 1,
                "height": 50,
                "justifyContent": "center",
                "margin": 2,
                "opacity": 1,
                "width": 50,
              }
            }
            testID="square"
          >
            <Text />
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
